### PR TITLE
Enforce naming rules on indexes and collections

### PR DESCRIPTION
**Pull request by scottinet**
# Description

Enforce strict naming rules on indexes and collections. This has the following effects:

* index and collection creation now return dedicated error messages/ids (`index:create` and `collection:create`)
* it's no longer possible to restrict a profile to an index or a collection that cannot possibly exist ([opened issue: KZL-1392](https://jira.kaliop.net/browse/KZL-1392))

The enforced rules are a stricter version of the [Elasticsearch naming rules](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/indices-create-index.html), as Kuzzle adds a few restrictions of its own.

The documentation has also been updated with these naming rules.

# Architecture

Naming rules have been implemented directly in the ES service, and propagated through the ClientAdapter class. I've made that choice because, to me, naming rules depend on the implemented storage service and therefore might change depending on it.

That way, if we use another storage service, it simply needs to implement the new `isIndexNameValid` and `isCollectionNameValid`.

This choice also explains why we have 1 dedicated function per object type (index & collection) even if this is useless with ES since we emulate collections with indexes. This paves the way for other storage engines allowing different structures and maybe different naming rules for indexes and collections.

### CODE REVIEW

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1682](https://github.com/kuzzleio/kuzzle/blob/583a26aec3701586a6d2b93bcea22c5469c038ba/lib/services/elasticsearch.js#L1682): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1736](https://github.com/kuzzleio/kuzzle/blob/583a26aec3701586a6d2b93bcea22c5469c038ba/lib/services/elasticsearch.js#L1736): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1816](https://github.com/kuzzleio/kuzzle/blob/583a26aec3701586a6d2b93bcea22c5469c038ba/lib/services/elasticsearch.js#L1816): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1682](https://github.com/kuzzleio/kuzzle/blob/5bf97a66bed57d23411f03ad40059b152d459551/lib/services/elasticsearch.js#L1682): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1736](https://github.com/kuzzleio/kuzzle/blob/5bf97a66bed57d23411f03ad40059b152d459551/lib/services/elasticsearch.js#L1736): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1816](https://github.com/kuzzleio/kuzzle/blob/5bf97a66bed57d23411f03ad40059b152d459551/lib/services/elasticsearch.js#L1816): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/services/elasticsearch.js [hunk 1]
```diff
@@ -746,34 +746,45 @@ class ElasticSearch extends Service {
...
 
-    return this._client.cat.indices({ format: 'json' })
-      .then(({ body }) => {
-        const esIndexes = body.map(({ index: name }) => name);
+    let body;
 
-        for (const esIndex of esIndexes) {
-          const indexName = this._extractIndex(esIndex);
+    try {
+      const res = await this._client.cat.indices({ format: 'json' });
```
Comment (Aschen) [inline lib/services/elasticsearch.js]:
```suggestion
      { body } = await this._client.cat.indices({ format: 'json' });
```
You can use destructuring assignment

File: lib/services/elasticsearch.js [hunk 2]
```diff
@@ -2108,11 +2126,48 @@ function assertWellFormedRefresh(esRequest) {
...
+  if (name.toLowerCase() !== name) {
+    return false;
+  }
+
+  if (Buffer.from(name).length > 126) {
+    return false;
+  }
+
+  let valid = true;
+  const forbiddenChars = `\\/*?"<>| \t\r\n,#:${NAME_SEPARATOR}${PUBLIC_PREFIX}${INTERNAL_PREFIX}`;
```
Comment (Aschen) [inline lib/services/elasticsearch.js]: (Nit): Can't we declare the forbidden chars once time only?

File: lib/services/elasticsearch.js (hunk 1 above)
Comment (alexandrebouthinon) [reply lib/services/elasticsearch.js]: Less weird syntax:
```suggestion
      const { body } = (await this._client.cat.indices({ format: 'json' }));
```

Comment (scottinet) [reply lib/services/elasticsearch.js]: @Aschen > dumb syntax is dumb :astonished: (I didn't know that :+1: )

@alexandrebouthinon > I can't do that (Dave), I need the body variable outside the try..catch

File: lib/services/elasticsearch.js [hunk 3]
```diff
@@ -746,34 +746,45 @@ class ElasticSearch extends Service {
...
 
-    return this._client.cat.indices({ format: 'json' })
-      .then(({ body }) => {
-        const esIndexes = body.map(({ index: name }) => name);
+    let body;
 
-        for (const esIndex of esIndexes) {
-          const indexName = this._extractIndex(esIndex);
+    try {
+      ({ body } = await this._client.cat.indices({ format: 'json' }));
```
Comment (kuzzle) [inline lib/services/elasticsearch.js]: MAJOR Extract the assignment of "{ body }" from this expression. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AAssignmentWithinCondition)

**Comment by kuzzle** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1682](https://github.com/kuzzleio/kuzzle/blob/08afc11375e238fb1df4ea328308bf8a64cdc51f/lib/services/elasticsearch.js#L1682): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1736](https://github.com/kuzzleio/kuzzle/blob/08afc11375e238fb1df4ea328308bf8a64cdc51f/lib/services/elasticsearch.js#L1736): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1816](https://github.com/kuzzleio/kuzzle/blob/08afc11375e238fb1df4ea328308bf8a64cdc51f/lib/services/elasticsearch.js#L1816): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1681](https://github.com/kuzzleio/kuzzle/blob/8857ca09a0c438304dae6fc0a4a319a9dd15cc4b/lib/services/elasticsearch.js#L1681): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1735](https://github.com/kuzzleio/kuzzle/blob/8857ca09a0c438304dae6fc0a4a319a9dd15cc4b/lib/services/elasticsearch.js#L1735): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1815](https://github.com/kuzzleio/kuzzle/blob/8857ca09a0c438304dae6fc0a4a319a9dd15cc4b/lib/services/elasticsearch.js#L1815): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1682](https://github.com/kuzzleio/kuzzle/blob/a53e36afaadf0f805972fd5140410dee3b5bb8ac/lib/services/elasticsearch.js#L1682): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1736](https://github.com/kuzzleio/kuzzle/blob/a53e36afaadf0f805972fd5140410dee3b5bb8ac/lib/services/elasticsearch.js#L1736): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1816](https://github.com/kuzzleio/kuzzle/blob/a53e36afaadf0f805972fd5140410dee3b5bb8ac/lib/services/elasticsearch.js#L1816): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1682](https://github.com/kuzzleio/kuzzle/blob/6013ffb61fe150dd8ea79cca2bcb37b331be17d2/lib/services/elasticsearch.js#L1682): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1736](https://github.com/kuzzleio/kuzzle/blob/6013ffb61fe150dd8ea79cca2bcb37b331be17d2/lib/services/elasticsearch.js#L1736): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1816](https://github.com/kuzzleio/kuzzle/blob/6013ffb61fe150dd8ea79cca2bcb37b331be17d2/lib/services/elasticsearch.js#L1816): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**